sound/soc/sof/ipc4-pcm.c

```diff
@@ -314,9 +315,99 @@ static int sof_ipc4_pcm_hw_params(struct snd_soc_component *component,
+					    struct snd_pcm_substream *substream,
+					    struct snd_sof_pcm_stream *stream,
+					    struct sof_ipc4_timestamp_info *time_info)
+{
+	struct sof_ipc4_copier *host_copier = time_info->host_copier;
+	struct sof_ipc4_copier *dai_copier = time_info->dai_copier;
+	struct sof_ipc4_pipeline_registers ppl_reg;
+	u64 stream_start_position;
+	u32 dai_sample_size;
+	u32 ch, node_id;
+	u32 offset;
+
+	node_id = SOF_IPC4_NODE_INDEX(host_copier->data.gtw_cfg.node_id);
+	offset = offsetof(struct sof_ipc4_fw_registers, pipeline_regs) + node_id * sizeof(ppl_reg);
+	sof_mailbox_read(sdev, sdev->fw_info_box.offset + offset, &ppl_reg, sizeof(ppl_reg));
+	if (ppl_reg.stream_start_offset == SOF_IPC4_INVALID_STREAM_POSITION)
+		return -EBUSY;
+
+	stream_start_position = ppl_reg.stream_start_offset;
+	ch = dai_copier->data.out_format.fmt_cfg;
+	ch = SOF_IPC4_AUDIO_FORMAT_CFG_CHANNELS_COUNT(ch);
+	dai_sample_size = (dai_copier->data.out_format.bit_depth >> 3) * ch;
+	/* convert offset to sample count */
+	do_div(stream_start_position, dai_sample_size);
+	time_info->stream_start_offset = stream_start_position;
+
+	return 0;
+}
+
+static snd_pcm_sframes_t sof_ipc4_pcm_delay(struct snd_soc_component *component,
+					    struct snd_pcm_substream *substream)
+{
+	struct snd_sof_dev *sdev = snd_soc_component_get_drvdata(component);
+	struct snd_soc_pcm_runtime *rtd = asoc_substream_to_rtd(substream);
+	struct sof_ipc4_timestamp_info *time_info;
+	struct sof_ipc4_llp_reading_slot llp;
+	struct snd_sof_pcm_stream *stream;
+	struct snd_sof_pcm *spcm;
+	snd_pcm_uframes_t host_ptr;
+	u64 link_ptr, tmp_ptr;
+	int ret;
+
+	spcm = snd_sof_find_spcm_dai(component, rtd);
+	if (!spcm)
+		return 0;
+
+	stream = &spcm->stream[substream->stream];
+	time_info = stream->private;
+	if (!time_info)
+		return 0;
+
+	/*
+	 * stream_start_offset is updated to memory window by FW based on
+	 * pipeline statistics and it may be invalid if host query happens before
+	 * the statistics is complete. And it will not change after the first initiailization.
+	 */
+	if (time_info->stream_start_offset == SOF_IPC4_INVALID_STREAM_POSITION) {
+		ret = sof_ipc4_get_stream_start_offset(sdev, substream, stream, time_info);
+		if (ret < 0)
+			return 0;
+	}
+
+	/* Intel HDA platform doesn't support llp counter based on GP-DMA */
+	if (!time_info->llp_offset) {
+		tmp_ptr = snd_sof_pcm_get_stream_position(sdev, component, substream);
+		if (!tmp_ptr)
+			return 0;
+	} else {
+		sof_mailbox_read(sdev, time_info->llp_offset, &llp, sizeof(llp));
+		tmp_ptr = ((u64)llp.reading.llp_l | ((u64)llp.reading.llp_u << 32));
+	}
+
+	/*
+	 * in rare case audio stream can last more than 24.855 hours
+	 * for 48khz 2ch case, do wrap to a 32 bit counter
+	 */
+	if (tmp_ptr > UINT_MAX) {
+		tmp_ptr -= 1;
+		link_ptr = do_div(tmp_ptr, UINT_MAX);
+	} else {
+		link_ptr = tmp_ptr;
+	}
+
+	host_ptr = substream->runtime->status->hw_ptr;
+
+	return host_ptr - (link_ptr - time_info->stream_start_offset);
```

Comment:
is this always a positive value? Should there be more range checks here?

Reply:
I would actually subtract the stream_start_offset from link_ptr **BEFORE** dealing with the wrap-around?

Reply:
How about capture, do we need to use   (link_ptr - time_info->stream_start_offset) - host_ptr ?

Reply:
@plbossart  I add some log and stack_dump to check the delay value. And find that two cases to call delay function
(1) called by snd_pcm_period_elapsed. At this time runtime->hw_ptr is not update and it is less than link_ptr, so delay is negative
(2) called by ioctl by user mode. runtime->hw_ptr is updated by last snd_pcm_period_elapsed and it is bigger than link_ptr, so delay is positive

In my opinion, the (2) is useful
````
[12132.297227] (NULL device *): delay -3698, host 36960, link 40658  // host - link < 0
[12132.297261] CPU: 5 PID: 2600 Comm: kworker/5:0 Tainted: G           O       6.0.0-rc5+ #760
[12132.297289] Hardware name: Dell Inc. XPS 9315/, BIOS 0.0.16 11/12/2021
[12132.297299] Workqueue: events snd_sof_pcm_period_elapsed_work [snd_sof]
[12132.297369] Call Trace:
[12132.297381]  <TASK>
[12132.297392]  dump_stack_lvl+0x33/0x46
[12132.297420]  sof_ipc4_pcm_delay+0x1cd/0x1e0 [snd_sof]
[12132.297501]  snd_soc_pcm_component_delay+0x51/0xf0 [snd_soc_core]
[12132.297603]  soc_pcm_pointer+0x5a/0x90 [snd_soc_core]
[12132.297674]  snd_pcm_update_hw_ptr0+0x43/0x400 [snd_pcm]
[12132.297720]  ? _snd_pcm_stream_lock_irqsave+0x26/0x30 [snd_pcm]
[12132.297768]  snd_pcm_period_elapsed_under_stream_lock+0x4c/0x80 [snd_pcm]
[12132.297803]  snd_pcm_period_elapsed+0x1a/0x30 [snd_pcm]
[12132.297836]  process_one_work+0x2c7/0x5d0
[12132.297873]  worker_thread+0x34/0x3d0
[12132.297892]  ? process_one_work+0x5d0/0x5d0
[12132.297904]  kthread+0xed/0x120
[12132.297915]  ? kthread_complete_and_exit+0x20/0x20
[12132.297932]  ret_from_fork+0x1f/0x30
[12132.297973]  </TASK>

[12132.298920] (NULL device *): delay 286, host 40992, link 40706 // host - link > 0
[12132.298946] CPU: 2 PID: 3125 Comm: speaker-test Tainted: G           O       6.0.0-rc5+ #760
[12132.298975] Hardware name: Dell Inc. XPS 9315/, BIOS 0.0.16 11/12/2021
[12132.298982] Call Trace:
[12132.298993]  <TASK>
[12132.299005]  dump_stack_lvl+0x33/0x46
[12132.299045]  sof_ipc4_pcm_delay+0x1cd/0x1e0 [snd_sof]
[12132.299146]  snd_soc_pcm_component_delay+0x51/0xf0 [snd_soc_core]
[12132.299252]  soc_pcm_pointer+0x5a/0x90 [snd_soc_core]
[12132.299332]  ? snd_pcm_hw_constraint_mask64+0x60/0x60 [snd_pcm]
[12132.299387]  snd_pcm_update_hw_ptr0+0x43/0x400 [snd_pcm]
[12132.299427]  ? __snd_pcm_lib_xfer+0x11b/0x8b0 [snd_pcm]
[12132.299465]  ? sched_clock_cpu+0xd/0xd0
[12132.299496]  ? snd_pcm_hw_constraint_mask64+0x60/0x60 [snd_pcm]
[12132.299536]  __snd_pcm_lib_xfer+0x81a/0x8b0 [snd_pcm]
[12132.299575]  ? _raw_spin_unlock_irqrestore+0x31/0x50
[12132.299596]  ? snd_pcm_hw_rule_step+0xe0/0xe0 [snd_pcm]
[12132.299632]  ? __wake_up_common_lock+0x82/0xb0
[12132.299669]  snd_pcm_common_ioctl+0x4b0/0x1010 [snd_pcm]
[12132.299711]  ? ioctl_has_perm.constprop.56+0xd6/0x140
[12132.299757]  snd_pcm_ioctl+0x25/0x30 [snd_pcm]
[12132.299791]  __x64_sys_ioctl+0x8a/0xd0
[12132.299822]  do_syscall_64+0x37/0x90
[12132.299845]  entry_SYSCALL_64_after_hwframe+0x63/0xcd
````